rmw_freertps_cpp/test/test_serialization.cpp

```diff
@@ -0,0 +1,615 @@
+    rmw_freertps_cpp::msg::UnboundedArrayBounded msg;
+    ASSERT_TRUE(rmw_freertps_cpp::msg::typesupport_freertps_cpp::deserialize_ros_msg(
+        &tmp_buf, &msg, &length));
+    for (uint32_t i = 0; i < 3; ++i) {
+      validate_factory_message(msg.primitive_values[i]);
+    }
+  }
+
+  free(buf);
+}
+
+TEST(test_serialization, serialize_unbounded_array_static) {
+  uint8_t * buf = nullptr;
+  uint32_t max_size = 0;
+  {
+    rmw_freertps_cpp::msg::UnboundedArrayStatic msg;
+    msg.primitive_values.resize(3);
+    for (uint32_t i = 0; i < 3; ++i) {
+      message_factory(msg.primitive_values[i]);
+    }
+    max_size = rmw_freertps_cpp::msg::typesupport_freertps_cpp::allocate_buffer_for_ros_msg(
+      &msg, buf);
+    uint32_t result = rmw_freertps_cpp::msg::typesupport_freertps_cpp::serialize_ros_msg(&msg, buf,
+        max_size);
+    EXPECT_GE(max_size, result);
+  }
+  {
+    uint32_t length = max_size;
+    uint8_t * tmp_buf = buf;
+    rmw_freertps_cpp::msg::UnboundedArrayStatic msg;
+    ASSERT_TRUE(rmw_freertps_cpp::msg::typesupport_freertps_cpp::deserialize_ros_msg(
+        &tmp_buf, &msg, &length));
+    for (uint32_t i = 0; i < 3; ++i) {
+      validate_factory_message(msg.primitive_values[i]);
+    }
+  }
+
+  free(buf);
+}
+
+TEST(test_serialization, serialize_unbounded_array_unbounded) {
+  uint8_t * buf = nullptr;
+  uint32_t max_size = 0;
+  {
+    rmw_freertps_cpp::msg::UnboundedArrayUnbounded msg;
+    msg.primitive_values.resize(3);
+    for (uint32_t i = 0; i < 3; ++i) {
+      message_factory(msg.primitive_values[i]);
+    }
+    max_size = rmw_freertps_cpp::msg::typesupport_freertps_cpp::allocate_buffer_for_ros_msg(
+      &msg, buf);
+    uint32_t result = rmw_freertps_cpp::msg::typesupport_freertps_cpp::serialize_ros_msg(&msg, buf,
+        max_size);
+    EXPECT_GE(max_size, result);
+  }
+  {
+    uint32_t length = max_size;
+    uint8_t * tmp_buf = buf;
+    rmw_freertps_cpp::msg::UnboundedArrayUnbounded msg;
+    ASSERT_TRUE(rmw_freertps_cpp::msg::typesupport_freertps_cpp::deserialize_ros_msg(
+        &tmp_buf, &msg, &length));
+    for (uint32_t i = 0; i < 3; ++i) {
+      validate_factory_message(msg.primitive_values[i]);
+    }
+  }
+
+  free(buf);
+}
+
+TEST(test_serialization, serialize_large_array) {
+  uint8_t * buf = nullptr;
+  uint32_t max_size = 0;
+  {
+    rmw_freertps_cpp::msg::LargeArray msg;
+    message_factory(msg);
+    max_size = rmw_freertps_cpp::msg::typesupport_freertps_cpp::allocate_buffer_for_ros_msg(
+      &msg, buf);
+    uint32_t result = rmw_freertps_cpp::msg::typesupport_freertps_cpp::serialize_ros_msg(&msg, buf,
+        max_size);
+    EXPECT_GE(max_size, result);
+  }
+  {
+    uint32_t length = max_size;
+    uint8_t * tmp_buf = buf;
+    rmw_freertps_cpp::msg::LargeArray msg;
+    ASSERT_TRUE(rmw_freertps_cpp::msg::typesupport_freertps_cpp::deserialize_ros_msg(
+        &tmp_buf, &msg, &length));
+    validate_factory_message(msg);
+  }
+
+  free(buf);
+}
```

Comment:
This file seems to have a lot of redundant code. Can some of it be refactored into functions / templates / macros to avoid the numerous repetitions?